### PR TITLE
Volume binding

### PR DESCRIPTION
**Pull request by KennethNagin**
Volume binding:
-- added support for volume binding, e.g.  create container -v
-- added volume management support:
    -- volume ls
    -- volume inspect
    -- volume rm
    -- volume create
-- bats tests for the above
note:
    This version only includes suffixing of tenant id to volume name  to accomplish both name scoping in tenant isolation.   Volume labels are not used to ear mark volumes as belonging to particular tenant.  Labels are not used since they are only available in newer releases ( release 1.11.0 and above) and label might not be available for volume binding in create container (needs investigation).  However since volumes are only identified by their name (volumes do not have IDs) appending the tenant id does provide tenant isolation.  Also a minor point we only allow volume creation when the user supplies a volume name, i.e. we don't support the case where docker generates a volume name.

### CODE REVIEW

**Comment by ezrasilvera** [comment]
Please rebase

File: pkg/multiTenancyPlugins/authorization/defaultAuthZImpl.go [hunk 1]
```diff
@@ -48,6 +48,10 @@ func (defaultauthZ *DefaultAuthZImpl) Handle(command utils.CommandEnum, cluster
 			if err := NetworkAuthorization(cluster, r, containerConfig.HostConfig.NetworkMode); err != nil {
 				return err
 			}
+			if err := hostFSMountCheck(r.Header.Get(headers.AuthZTenantIdHeaderName), containerConfig.HostConfig.Binds); err != nil {
+				return err
+			}
+
```
Comment (ezrasilvera) [inline pkg/multiTenancyPlugins/authorization/defaultAuthZImpl.go]: Should this be moved to the argument enforcement ?

Comment (KennethNagin) [inline pkg/multiTenancyPlugins/authorization/defaultAuthZImpl.go]: > From: Ezra Silvera notifications@github.com
> To: swarm-hooks/swarm swarm@noreply.github.com
> Cc: Kenneth Nagin/Haifa/IBM@IBMIL, Author author@noreply.github.com
> Date: 19/07/2016 04:24 PM
> Subject: Re: [swarm-hooks/swarm] Volume binding (#46)
> 
> In pkg/multiTenancyPlugins/authorization/defaultAuthZImpl.go:
> 
> > @@ -48,6 +48,10 @@ func (defaultauthZ *DefaultAuthZImpl) Handle
> > (command utils.CommandEnum, cluster
> >           if err := NetworkAuthorization(cluster, r, 
> > containerConfig.HostConfig.NetworkMode); err != nil {
> >              return err
> >           }
> > -         if err := hostFSMountCheck(r.Header.Get
> >   (headers.AuthZTenantIdHeaderName), 
> >   containerConfig.HostConfig.Binds); err != nil {
> > -            return err
> > -         }
> >   +
> 
> Should this be moved to the argument enforcement ? 
> ?
> Indeed, if we had some other "argument enforcement" code this would be a 
> candidate.
> However, no such code exists nor do we have a design whether we want it 
> fit the model
> of api filtering.  Once we understand how "argument enforcement" will be 
> done
> then we can decide whether to include -v argument handling.
> For now I think it is best to include it in the authorization plugin.

File: pkg/multiTenancyPlugins/authorization/volumes.go [hunk 2]
```diff
@@ -40,21 +38,21 @@ func filterVolumes(r *http.Request, rec *httptest.ResponseRecorder) []byte {
...
-	for _,b := range volumeBindings {
-		if index := strings.Index(b,":"); index > -1 {
-		  v := b[0:index]
-		  if strings.Contains(v,"/") {
-			return errors.New(HOST_FS_MOUNT_NOT_ALLOWED)
-		  }
+func hostFSMountCheck(tenantId string, volumeBindings []string) error {
+	for _, b := range volumeBindings {
+		if index := strings.Index(b, ":"); index > -1 {
+			v := b[0:index]
```
Comment (doronp) [inline pkg/multiTenancyPlugins/authorization/volumes.go]: Can it be done using regexp? IMHO using indexes like that is not so readable and is more error prone.

File: pkg/multiTenancyPlugins/naming/volumes.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		r, _ = utils.ModifyRequest(r, bytes.NewReader(buf.Bytes()), "", "")
+	}
+	return nil
+}
+// append tenantid to volume name
+func updateVolumeName(r *http.Request, muxVolumeName string) {
+	volumeName := mux.Vars(r)[muxVolumeName] + r.Header.Get(headers.AuthZTenantIdHeaderName)
+	r.URL.Path = strings.Replace(r.URL.Path, mux.Vars(r)[muxVolumeName], volumeName, 1)
+	mux.Vars(r)[muxVolumeName] = volumeName
+	return
```
Comment (doronp) [inline pkg/multiTenancyPlugins/naming/volumes.go]: Remove this line

File: pkg/multiTenancyPlugins/naming/volumes.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+package namescoping
+
+import (
+	log "github.com/Sirupsen/logrus"
+	"github.com/samalba/dockerclient"
+	"errors"
+	"strings"
+	"bytes"
+	"encoding/json"
+	apitypes "github.com/docker/engine-api/types"
```
Comment (doronp) [inline pkg/multiTenancyPlugins/naming/volumes.go]: Please use the new apiTypes

File: pkg/multiTenancyPlugins/naming/volumes.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
+// volumes
+package namescoping
+
+import (
+	log "github.com/Sirupsen/logrus"
+	"github.com/samalba/dockerclient"
```
Comment (doronp) [inline pkg/multiTenancyPlugins/naming/volumes.go]: Please use the new API types and remove the old dockerClient

File: pkg/multiTenancyPlugins/authorization/volumes.go [hunk 6]
```diff
@@ -2,34 +2,32 @@
...
 	if err := json.NewDecoder(bytes.NewReader(rec.Body.Bytes())).Decode(&response); err != nil {
 		log.Error(err)
 		return nil
 	}
 	var candidates []*apitypes.Volume
 	tenantId := r.Header.Get(headers.AuthZTenantIdHeaderName)
 	for _, volume := range response.Volumes {
-		if strings.HasSuffix(volume.Name,tenantId) {
-			candidates = append(candidates, volume)			
+		if strings.HasSuffix(volume.Name, tenantId) {
```
Comment (doronp) [inline pkg/multiTenancyPlugins/authorization/volumes.go]: I think in containers we agreed using label for that purpose. Can this be done here as well? Having two methods of achieving similar thing is confusing.

File: pkg/multiTenancyPlugins/naming/nameScoping.go [hunk 7]
```diff
@@ -14,8 +14,8 @@ import (
 	"github.com/samalba/dockerclient"
```
Comment (doronp) [inline pkg/multiTenancyPlugins/naming/nameScoping.go]: Again - I see the use of new client - Please use only the new one

File: pkg/multiTenancyPlugins/authorization/volumes.go [hunk 8]
```diff
@@ -2,34 +2,32 @@
...
 
 func filterVolumes(r *http.Request, rec *httptest.ResponseRecorder) []byte {
-	var response apitypes.VolumesListResponse 
+	var response apitypes.VolumesListResponse
 	if err := json.NewDecoder(bytes.NewReader(rec.Body.Bytes())).Decode(&response); err != nil {
 		log.Error(err)
 		return nil
 	}
 	var candidates []*apitypes.Volume
 	tenantId := r.Header.Get(headers.AuthZTenantIdHeaderName)
```
Comment (doronp) [inline pkg/multiTenancyPlugins/authorization/volumes.go]: IMHO instead of putting the value in a var just put it directly to the 
`if strings.HasSuffix(volume.Name, tenantId)`

File: pkg/multiTenancyPlugins/authorization/volumes.go [hunk 9]
```diff
@@ -40,21 +38,21 @@ func filterVolumes(r *http.Request, rec *httptest.ResponseRecorder) []byte {
...
-func hostFSMountCheck(tenantId string, volumeBindings[] string) (error) {
-	for _,b := range volumeBindings {
-		if index := strings.Index(b,":"); index > -1 {
-		  v := b[0:index]
-		  if strings.Contains(v,"/") {
-			return errors.New(HOST_FS_MOUNT_NOT_ALLOWED)
-		  }
+func hostFSMountCheck(tenantId string, volumeBindings []string) error {
+	for _, b := range volumeBindings {
+		if index := strings.Index(b, ":"); index > -1 {
```
Comment (doronp) [inline pkg/multiTenancyPlugins/authorization/volumes.go]: See comment on the other use of this method - IMHO use regexp if possible

File: pkg/multiTenancyPlugins/authorization/volumes.go [hunk 10]
```diff
@@ -40,21 +38,21 @@ func filterVolumes(r *http.Request, rec *httptest.ResponseRecorder) []byte {
...
+			v := b[0:index]
+			if strings.Contains(v, "/") {
```
Comment (doronp) [inline pkg/multiTenancyPlugins/authorization/volumes.go]: Does that cover all cases? What if there is something like `dir1/dir2`

**Comment by doronp** [comment]
@KennethNagin @ezrasilvera I've reviewed:
Not LGTM. Please see comments

File: pkg/multiTenancyPlugins/authorization/volumes.go (hunk 10 above)
Comment (KennethNagin) [inline pkg/multiTenancyPlugins/authorization/volumes.go]: > From: Doron Podoleanu notifications@github.com
> To: swarm-hooks/swarm swarm@noreply.github.com
> Cc: Kenneth Nagin/Haifa/IBM@IBMIL, Author author@noreply.github.com
> Date: 20/07/2016 10:46 AM
> Subject: Re: [swarm-hooks/swarm] Volume binding (#46)
> 
> In pkg/multiTenancyPlugins/authorization/volumes.go:
> 
> > @@ -40,21 +38,21 @@ func filterVolumes(r *http.Request, rec 
> > *httptest.ResponseRecorder) []byte {
> >     }
> >     return buf.Bytes()
> >  }
> > -func hostFSMountCheck(tenantId string, volumeBindings[] string) 
> > (error) {
> > -   for _,b := range volumeBindings {
> > -      if index := strings.Index(b,":"); index > -1 {
> > -        v := b[0:index]
> > -        if strings.Contains(v,"/") {
> > -         return errors.New(HOST_FS_MOUNT_NOT_ALLOWED)
> > -        }
> >   +func hostFSMountCheck(tenantId string, volumeBindings []string) error 
> >   {
> > -   for _, b := range volumeBindings {
> > -      if index := strings.Index(b, ":"); index > -1 {
> > -         v := b[0:index]
> > -         if strings.Contains(v, "/") {
> 
> Does that cover all cases? What if there is something like dir1/dir2
> dir1/dir2 is an illegal combination:
> nagin@rcc-hrl-kvg-558:~$ sudo docker run  -v tmp/flavors:/tmp busybox top
> docker: Error response from daemon: create tmp/flavors: volume name 
> invalid: "tmp/flavors" includes invalid characters for a local volume 
> name, only "[a-zA-Z0-9][a-zA-Z0-9_.-]" are allowed.
> See 'docker run --help'.

File: pkg/multiTenancyPlugins/authorization/volumes.go (hunk 9 above)
Comment (KennethNagin) [inline pkg/multiTenancyPlugins/authorization/volumes.go]: > From: Doron Podoleanu notifications@github.com
> To: swarm-hooks/swarm swarm@noreply.github.com
> Cc: Kenneth Nagin/Haifa/IBM@IBMIL, Author author@noreply.github.com
> Date: 20/07/2016 10:45 AM
> Subject: Re: [swarm-hooks/swarm] Volume binding (#46)
> 
> In pkg/multiTenancyPlugins/authorization/volumes.go:
> 
> > @@ -40,21 +38,21 @@ func filterVolumes(r *http.Request, rec 
> > *httptest.ResponseRecorder) []byte {
> >     }
> >     return buf.Bytes()
> >  }
> > -func hostFSMountCheck(tenantId string, volumeBindings[] string) 
> > (error) {
> > -   for _,b := range volumeBindings {
> > -      if index := strings.Index(b,":"); index > -1 {
> > -        v := b[0:index]
> > -        if strings.Contains(v,"/") {
> > -         return errors.New(HOST_FS_MOUNT_NOT_ALLOWED)
> > -        }
> >   +func hostFSMountCheck(tenantId string, volumeBindings []string) error 
> >   {
> > -   for _, b := range volumeBindings {
> > -      if index := strings.Index(b, ":"); index > -1 {
> 
> See comment on the other use of this method - IMHO use regexp if 
> possible
> Personally,  I feel the usage on regexp in this case would be more 
> confusing.  Ezra what is your preference?
> --Ken

File: pkg/multiTenancyPlugins/naming/nameScoping.go (hunk 7 above)
Comment (KennethNagin) [inline pkg/multiTenancyPlugins/naming/nameScoping.go]: The branch was about adding volume support, not about replacing the use of "github.com/samalba/dockerclient" with apitype in create container.  I simply followed the pattern that was already in place in create container to also support the -v flag.
I suggest creating a separate pull request to replacing  "github.com/samalba/dockerclient" with apitype in create container and not mix apples with oranges.
--Ken

File: pkg/multiTenancyPlugins/authorization/volumes.go (hunk 2 above)
Comment (KennethNagin) [inline pkg/multiTenancyPlugins/authorization/volumes.go]: regexp is very powerful, but also very very error prone.  Ezra: what do you prefer?

File: pkg/multiTenancyPlugins/naming/volumes.go (hunk 3 above)
Comment (KennethNagin) [inline pkg/multiTenancyPlugins/naming/volumes.go]: Which line is bothering you?

File: pkg/multiTenancyPlugins/naming/volumes.go (hunk 4 above)
Comment (KennethNagin) [inline pkg/multiTenancyPlugins/naming/volumes.go]: Same comment as above.  "github.com/samalba/dockerclient" is used here because it is used in create container.  I think the change to create container should be handled globally in a separate pull request.

File: pkg/multiTenancyPlugins/authorization/volumes.go (hunk 6 above)
Comment (KennethNagin) [inline pkg/multiTenancyPlugins/authorization/volumes.go]: There was a long thread on this subject in hrl-swarm.  You made no comment. Ezra decided against using volume labels.

File: pkg/multiTenancyPlugins/naming/nameScoping.go (hunk 7 above)
Comment (doronp) [inline pkg/multiTenancyPlugins/naming/nameScoping.go]: Agree so I think it is best (to worst):
1 - First Do the engine-api refactor and avoid introducing new code which still uses the old client, only then add the volumes code.
2 - remove the unused import from the imports and continue using the old client
3 - Do nothing at all

Comment (ezrasilvera) [inline pkg/multiTenancyPlugins/naming/nameScoping.go]: @KennethNagin To clarify - You should avoid using dockerclient structures in YOUR code even if it is already beaning used in that file.  I understand that using engine-api/types in YOUR code may require you to change existing (not yours) code as well.  But it's not reasonable to add more code that someone else later on will need to change as well. 
It's OK that you make changes that are related only to structures your are using .

File: pkg/multiTenancyPlugins/authorization/volumes.go (hunk 10 above)
Comment (ezrasilvera) [inline pkg/multiTenancyPlugins/authorization/volumes.go]: Ken, why don';t you look in the engine code to see (and take if possible) the parsing ?

Comment (KennethNagin) [inline pkg/multiTenancyPlugins/authorization/volumes.go]: Interesting problem:
   There are two parsers, one in volume/volume_unix.go and one in volume/volume_windows.go, with a function ParseMountSpec.   volume/volume_unix.go uses filepath.ToSlash, strings.SplitN, and strings.Count to parse the bind.     volume/volume_windows.go use some complicated regex to do the job.

I have not figured out how the code decides which parser to use ( I think the call happens in daemon/volumes.go registerMountPoints).  However, from the names I'm guessing that when docker is running in unix then volume_unix.go called.

Should a modify the code to do similar parse to volume_unix.go?

File: pkg/multiTenancyPlugins/naming/nameScoping.go (hunk 7 above)
Comment (KennethNagin) [inline pkg/multiTenancyPlugins/naming/nameScoping.go]: @ezrasilvera  In order to minimize changes to code not related to volume_binding it will be necessary to add an extra encode/decode of the request  to/from engine-api/types just for volume_binding.  Is this what you want?

File: pkg/multiTenancyPlugins/naming/nameScoping.go [hunk 11]
```diff
@@ -248,3 +272,24 @@ func getContainerId(container *cluster.Container, tenantId string, containerRefe
 	}
 	return "", errors.New(CONTAINER_REFERENCE_NOT_FOR_CONTAINER_INFO)
 }
+
+// send request to next handler in record mode.  When it returns remove traces of tenant id in response
+func doReq(nameScoping *DefaultNameScopingImpl, command utils.CommandEnum, cluster cluster.Cluster, w http.ResponseWriter, r *http.Request, swarmHandler http.Handler) error {
+	newReq, e1 := utils.ModifyRequest(r, nil, "", "")
```
Comment (ezrasilvera) [inline pkg/multiTenancyPlugins/naming/nameScoping.go]: Why do you need a new req as well ?  What is the modification here ?

File: pkg/multiTenancyPlugins/naming/nameScoping.go [hunk 12]
```diff
@@ -248,3 +272,24 @@ func getContainerId(container *cluster.Container, tenantId string, containerRefe
 	}
 	return "", errors.New(CONTAINER_REFERENCE_NOT_FOR_CONTAINER_INFO)
 }
+
+// send request to next handler in record mode.  When it returns remove traces of tenant id in response
+func doReq(nameScoping *DefaultNameScopingImpl, command utils.CommandEnum, cluster cluster.Cluster, w http.ResponseWriter, r *http.Request, swarmHandler http.Handler) error {
```
Comment (ezrasilvera) [inline pkg/multiTenancyPlugins/naming/nameScoping.go]: Why not just do the manipulation in the function and create a recorder and send to nect handler from the main case statement.

Comment (KennethNagin) [inline pkg/multiTenancyPlugins/naming/nameScoping.go]: Ok.  I will move the encode/decode to/from apitypes.VolumeCreateRequest out of volumes.go to nameScoping.go.

File: pkg/multiTenancyPlugins/naming/nameScoping.go (hunk 11 above)
Comment (KennethNagin) [inline pkg/multiTenancyPlugins/naming/nameScoping.go]: This is a hold over from following Doron's modifyRequest utility to modify the request.
I will investigate to figure out how to slim down the create volume to its bear mininum.

File: pkg/multiTenancyPlugins/authorization/defaultAuthZImpl.go [hunk 13]
```diff
@@ -50,6 +50,9 @@ func (defaultauthZ *DefaultAuthZImpl) Handle(command utils.CommandEnum, cluster
 				return err
 			}
 			oldconfig.ContainerConfig.Config.Labels[headers.TenancyLabel] = r.Header.Get(headers.AuthZTenantIdHeaderName)
+			if err := hostFSMountCheck(r.Header.Get(headers.AuthZTenantIdHeaderName), oldconfig.HostConfig.Binds); err != nil {
```
Comment (ezrasilvera) [inline pkg/multiTenancyPlugins/authorization/defaultAuthZImpl.go]: let's remove this test for now as we allow many arguments that should be disabled for multi tennat. Let's move all of these checks to another place

**Comment by ezrasilvera** [comment]
please rebase and see the comment about the mount check
